middleware/http/oauth2/oauth2_middleware.go

```diff
@@ -15,144 +15,299 @@ package oauth2
+		//nolint:nilerr
+		return nil, nil
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to retrieve cookie: %w", err)
+	}
+
+	// Decrypt the encrypted (JWE) cookie
+	dec, err := jwe.Decrypt(
+		[]byte(cookie.Value),
+		jwe.WithKey(jwa.A128KW, m.meta.cek),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt cookie: %w", err)
+	}
+
+	// Validate the JWT from the decrypted cookie
+	token, err := jwt.Parse(dec,
+		jwt.WithKey(jwa.HS256, m.meta.csk),
+		jwt.WithIssuer(jwtIssuer),
+		jwt.WithAudience(m.meta.ClientID), // Use the client ID as audience
+		jwt.WithAcceptableSkew(allowedClockSkew),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate JWT token: %w", err)
+	}
+
+	return cast.ToStringMapString(token.PrivateClaims()), nil
 }
 
-func (m *Middleware) getNativeMetadata(metadata middleware.Metadata) (*oAuth2MiddlewareMetadata, error) {
-	var middlewareMetadata oAuth2MiddlewareMetadata
-	err := mdutils.DecodeMetadata(metadata.Properties, &middlewareMetadata)
+func (m *Middleware) setSecureCookie(w http.ResponseWriter, claims map[string]string, ttl time.Duration, domain string, isSecure bool) error {
+	now := time.Now()
+	exp := now.Add(ttl)
+
+	// Build the JWT
+	builder := jwt.NewBuilder().
+		IssuedAt(now).
+		Issuer(jwtIssuer).
+		Audience([]string{m.meta.ClientID}).
+		Expiration(exp).
+		NotBefore(now)
+	var claimsSize int
+	for k, v := range claims {
+		builder.Claim(k, v)
+		claimsSize += len(v)
+	}
+	token, err := builder.Build()
+	if err != nil {
+		return fmt.Errorf("error building JWT: %w", err)
+	}
+
+	// Generate the encrypted JWT
+	var encryptOpts []jwt.EncryptOption
+	if claimsSize > 800 {
+		// If the total size of the claims is more than 800 bytes, we should enable compression
+		encryptOpts = []jwt.EncryptOption{
+			jwt.WithKey(jwa.A128KW, m.meta.cek),
+			jwt.WithEncryptOption(jwe.WithContentEncryption(jwa.A128GCM)),
+			jwt.WithEncryptOption(jwe.WithCompress(jwa.Deflate)),
+		}
+	} else {
+		encryptOpts = []jwt.EncryptOption{
+			jwt.WithKey(jwa.A128KW, m.meta.cek),
+			jwt.WithEncryptOption(jwe.WithContentEncryption(jwa.A128GCM)),
+		}
+	}
+	val, err := jwt.NewSerializer().
+		Sign(
+			jwt.WithKey(jwa.HS256, m.meta.csk),
+		).
+		Encrypt(encryptOpts...).
+		Serialize(token)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to serialize token: %w", err)
 	}
-	return &middlewareMetadata, nil
+
+	// Generate the cookie
+	cookie := http.Cookie{
+		Name:     m.meta.CookieName,
+		Value:    string(val),
+		MaxAge:   int(ttl.Seconds()),
+		HttpOnly: true,
+		Secure:   isSecure,
+		Domain:   domain,
+		Path:     "/",
+		SameSite: http.SameSiteLaxMode,
+	}
+	cookieStr := cookie.String()
+
+	// Browsers have a maximum size of about 4KB, and if the cookie is larger than that (by looking at the entire value of the header), it is silently rejected
+	// Some info: https://stackoverflow.com/a/4604212/192024
+	if len(cookieStr) > 4<<10 {
```

Comment:
What happens if the backend service also adds cookie values on subsequent requests? Is the total limit still 4k?

Reply:
Yes, it's still 4KB per domain in most browsers. If the cookies go over the limit, they are silently dropped